### PR TITLE
UI Refactor: update state and draw in separate threads

### PR DESCRIPTION
**Pull request by deanlee**
1.mv all states to class UIScene
2.update states from another thread

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
This seems more complicated than it needs to be. If we can atomically read from a UI state struct in the drawing thread, then we can make this simpler.

**Comment by deanlee** [comment]
Did u mean passing a state struct by value?

**Comment by deanlee** [comment]
It’s still too complicated. I’ll think about how to make it better.

**Comment by adeebshihadeh** [comment]
I think if we do this right, it’ll be easier to read than before :)

**Comment by deanlee** [comment]
Done,I am satisfied with it personally.

**Comment by deanlee** [comment]
the state thread processes and copies data extremely fast.now the main bottleneck is in function visionstream_get().most of the "slow frame" messages are caused by it.I'm confused why is it laggy

**Comment by deanlee** [comment]
> the state thread processes and copies data extremely fast.now the main bottleneck is in function visionstream_get().most of the "slow frame" messages are caused by it.I'm confused why is it laggy

I now know why this happens, because of the wrong flag, poll did not work as expected. I submitted a PR to fix this: https://github.com/commaai/openpilot/pull/2550

**Comment by adeebshihadeh** [comment]
Nice work! Let's merge #2550 into this PR. I'll have some time to review this after 0.8 is released.

File: selfdrive/ui/ui.cc [hunk 1]
```diff
@@ -157,6 +157,22 @@ static void update_model(UIScene &scene, const cereal::ModelDataV2::Reader &mode
   update_line_data(scene, model.getPosition(), 0.5, 0, &scene.track_vertices, path_length);
 }
 
+static void update_lead(const cereal::RadarState::LeadData::Reader &lead,
```
Comment (deanlee) [inline selfdrive/ui/ui.cc:160]: @HaraldSchafer: This function is still in the draft stage, do you think it is correct?

Comment (HaraldSchafer) [reply selfdrive/ui/ui.cc]: Looks correct to me!

**Comment by deanlee** [comment]
refactor alert   https://github.com/commaai/openpilot/pull/2751/files

**Comment by deanlee** [comment]
convert to draft,the code of the qt part has been changing

**Comment by adeebshihadeh** [comment]
I like the direction of this, but it's a big PR. Can we break it up into smaller PRs and build up to it?

**Comment by deanlee** [comment]
How about using a PR to put all the data that needs to be synchronized into the struct UIScene? Then use another PR to complete the synchronization

**Comment by adeebshihadeh** [comment]
Hmm, that still sounds like big PRs. If this takes big PRs, I think it will sit around for a while, since the UI is stable and not a high priority.

**Comment by deanlee** [comment]
What about using variables instead of cereal::read to get data? If we want to update in the background, we can no longer use cereal::read

**Pull request closed without merging**